### ISSUE TITLE
Refactor chpl to act as a compiler-driver

### ISSUE THREAD

**Issue by riftEmber** [chapel-lang/chapel#20871]
The `chpl` compiler currently performs all stages of building Chapel code without releasing memory in-between; the compilation and assembly stages are performed as a monolithic process, followed by linking as a subprocess. This means more memory is used; for example, the entire AST from compilation is still unnecessarily present in memory during linking.

This proposal is to use a compiler-driver approach, where `chpl` itself is just a driver invoking each stage of building in sequence. The driver terminates each process once its task is finished (reclaiming memory) and passes its outputs as inputs to the next. `gcc` for example takes a compiler-driver approach, invoking in turn the preprocessor, compiler, assembler, and linker.

**Comment by vasslitvinov** [comment]
One feature that this approach will enable is the ability to switch configurations without rebuilding the compiler. For example the driver will choose the appropriate compiler executable depending on C vs . LLVM backend or whether CHPL_LLVM is enabled. Cf. today this switch requires re-linking the compiler executable.

**Comment by riftEmber** [comment]
As I pick this up, reposting some todo notes from the [last sprint task I had on this topic](https://github.com/Cray/chapel-private/issues/4160). These are refactors needed for the prototype to be production quality, but are not blocking progress yet.
- [ ] don't init as much stuff before splitting off in the driver since that work is redone later
- [ ] don't do runClang in makeBinary as it does more than needed -- extract needed functionality to another function, as attempted in https://github.com/chapel-lang/chapel/commit/cbc260b9ab9738ee7e41e602040dcf22b4098feb

**Comment by riftEmber** [comment]
With https://github.com/chapel-lang/chapel/pull/21063 merged, we now have `--compiler-driver` to opt-in to running `chpl` as a compiler-driver; some future work is tracked in https://github.com/Cray/chapel-private/issues/5373 which will culminate in using compiler-driver behavior by default with an opt-out option.

**Comment by riftEmber** [comment]
With https://github.com/chapel-lang/chapel/pull/23930 driver mode is now on by default, with the opt-out `--no-compiler-driver` provided.

**Comment by riftEmber** [comment]
Closing this as done since driver is now default. Eventually we might make it the only mode, but that transition is tracked in https://github.com/Cray/chapel-private/issues/5618.
